### PR TITLE
Strengthen validateClusterFormed check

### LINKED ISSUE #49243: [CI] Failure in NoMasterNodeIT.testNoMasterActions
This just failed in `7.x` on CI:  https://gradle-enterprise.elastic.co/s/546gdpwyjvnyc/console-log

```

2> REPRODUCE WITH: ./gradlew ':server:integTest' --tests "org.elasticsearch.cluster.NoMasterNodeIT.testNoMasterActions" -Dtests.seed=C5890400A905986B -Dtests.security.manager=true -Dtests.locale=ar-OM -Dtests.timezone=SystemV/YST9 -Dcompiler.java=12 -Druntime.java=8
--
2> java.lang.AssertionError: AcknowledgedResponse failed - not acked
Expected: <true>
but: was <false>
2> NOTE: leaving temporary files on disk at: /dev/shm/elastic+elasticsearch+7.x+multijob+fast+part1/server/build/testrun/integTest/temp/org.elasticsearch.cluster.NoMasterNodeIT_C5890400A905986B-001
2> NOTE: test params are: codec=Asserting(Lucene80), sim=Asserting(org.apache.lucene.search.similarities.AssertingSimilarity@55a614c0), locale=ar-OM, timezone=SystemV/YST9

```

Does not reproduce locally over a few 10s of runs at least.

### PR DESCRIPTION
**Pull request by ywelsch**
Strengthens the `validateClusterFormed` check that is used by the test infrastructure to make sure that nodes are properly connected and know about each other. Is used in situations where the cluster is scaled up and down, and where there previously was a network disruption that has been healed.

Closes #49243

### CODE REVIEW

File: test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java [hunk 1]
```diff
@@ -1092,29 +1093,36 @@ private synchronized void reset(boolean wipeData) throws IOException {
...
-                }
+                final List<ClusterState> states = stateSupplier.get();
+                // all nodes have a master
+                assertTrue(states.toString(), states.stream().allMatch(cs -> cs.nodes().getMasterNodeId() != null));
+                // all nodes have the same master (in same term)
+                assertEquals(states.toString(), 1, states.stream().mapToLong(ClusterState::term).distinct().count());
+                // all nodes know about all other nodes
+                states.forEach(cs -> {
+                    DiscoveryNodes discoveryNodes = cs.nodes();
+                    assertEquals(expectedNodes.size(), discoveryNodes.getSize());
```
Comment (DaveCTurner) [inline test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java]: Could you label this with `states.toString()` so we can see the mismatch on failure?

File: test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java [hunk 2]
```diff
@@ -1092,29 +1093,36 @@ private synchronized void reset(boolean wipeData) throws IOException {
...
+                    for (DiscoveryNode expectedNode : expectedNodes) {
+                        assertTrue("Expected node to exist: " + expectedNode, discoveryNodes.nodeExists(expectedNode));
```
Comment (DaveCTurner) [inline test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java]: Could you label this with `states.toString()` so we can see the mismatch on failure?

File: test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java [hunk 3]
```diff
@@ -1092,29 +1093,36 @@ private synchronized void reset(boolean wipeData) throws IOException {
...
+                    }
+                });
             }, 30, TimeUnit.SECONDS);
         } catch (AssertionError ae) {
-            throw new IllegalStateException("cluster failed to form with expected nodes " + expectedNodes + " and actual nodes " +
-                client.admin().cluster().prepareState().get().getState().nodes());
+            throw new IllegalStateException("cluster failed to form with expected nodes " + expectedNodes + " and actual cluster states " +
+                stateSupplier.get(), ae);
```
Comment (DaveCTurner) [inline test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java]: Do we really need to include the actual cluster states in this message too? I think it'll be in the `AssertionError`.

**Review by DaveCTurner — COMMENTED** [review]
I asked for some minor changes but mostly looks good.

**Comment by ywelsch** [comment]
I've addressed all comments. Can you have another look?

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Pull request merged**